module/surface_sam_sid_vhf.c

```diff
@@ -0,0 +1,411 @@
+	status = surface_sam_ssh_rqst(&rqst, &result);
+	if (status) {
+		return status;
+	}
+
+	len = resp.data.info.hid_len;
+
+	// allocate a buffer for the descriptor
+	buf = kzalloc(len, GFP_KERNEL);
+
+	// then, iterate and write into buffer, copying out resp.rqst.hid_len bytes
+	// (it gets set to the length of data read)
+
+	resp.rqst.id = 1;
+	resp.rqst.offset = 0;
+	resp.rqst.limit = 0x76;
+	resp.rqst.end = 0;
+
+	while (resp.rqst.end != 0x01 && resp.rqst.offset < len) {
+		status = surface_sam_ssh_rqst(&rqst, &result);
+		if (status) {
+			kfree(buf);
+			return status;
+		}
+		memcpy(buf + resp.rqst.offset, resp.data.pld, resp.rqst.limit);
+
+		resp.rqst.offset += resp.rqst.limit;
+	}
+
+	*desc = buf;
+	*size = len;
+
+	return 0;
+}
+
+static int sid_vhf_hid_parse(struct hid_device *hid)
+{
+	int ret = 0, size;
+	u8 *buf;
+
+	ret = vhf_get_hid_descriptor(hid, 0x00, &buf, &size);
+	if (ret != 0) {
+		hid_dbg(hid, "vhf_get_hid_descriptor ret %d\n", ret);
+		return -EIO;
+	}
+	print_hex_dump_debug("descriptor:", DUMP_PREFIX_OFFSET, 16, 1, buf, size, false);
+
+	ret = hid_parse_report(hid, buf, size);
+	kfree(buf);
+	return ret;
+
+}
+
+static int sid_vhf_hid_raw_request(struct hid_device *hid, unsigned char reportnum,
+			       u8 *buf, size_t len, unsigned char rtype,
+			       int reqtype)
+{
+	int status;
+	u8 cid;
+	struct surface_sam_ssh_rqst rqst = {};
+	struct surface_sam_ssh_buf result = {};
+
+	hid_dbg(hid, "%s: reportnum %#04x %i %i\n", __func__, reportnum, rtype, reqtype);
+	print_hex_dump_debug("report:", DUMP_PREFIX_OFFSET, 16, 1, buf, len, false);
+
+	// Byte 0 is the report number. Report data starts at byte 1.
+	buf[0] = reportnum;
+
+	switch (rtype) {
+	case HID_OUTPUT_REPORT:
+		cid = 0x01;
+		break;
+	case HID_FEATURE_REPORT:
+		switch (reqtype) {
+		case HID_REQ_GET_REPORT:
+			// The EC doesn't respond to GET FEATURE for these touchpad reports
+			// we immediately discard to avoid waiting for a timeout.
+			if (reportnum == 6 || reportnum == 7 || reportnum == 8 || reportnum == 9 || reportnum == 0x0b) {
+				hid_dbg(hid, "%s: skipping get feature report for 0x%02x\n", __func__, reportnum);
+				return 0;
+			}
+
+			cid = 0x02;
+			break;
+		case HID_REQ_SET_REPORT:
+			cid = 0x03;
+			break;
+		default:
+			hid_err(hid, "%s: unknown req type 0x%02x\n", __func__, rtype);
+			return 0;	// discard everything
+		}
+                break;
+	default:
+		hid_err(hid, "%s: unknown report type 0x%02x\n", __func__, reportnum);
+		return 0;		// discard everything
+	}
+
+        rqst.tc  = SAM_EVENT_SID_VHF_TC;
+        rqst.pri = SURFACE_SAM_PRIORITY_HIGH;
+        rqst.iid = 0x00; // windows tends to distinguish iids, but EC will take it
+        rqst.cid = cid;
+        rqst.snc = HID_REQ_GET_REPORT == reqtype ? 0x01 : 0x00;
+        rqst.cdl = HID_REQ_GET_REPORT == reqtype ? 0x01 : len;
+        rqst.pld = buf;
+
+        result.cap = len;
+        result.len = 0;
+        result.data = buf;
```

Comment:
There's something wrong with the indentation here.

Reply:
Yeah it's spaces, fixing